difftime.1
.TH DIFFTIME 1 2020-11-17 GNU "User Commands"
.SH NAME
difftime \- display delta time from two files or one file and now
.SH SYNOPSIS
.B difftime
[\fI\,OPTION\/\fR]... \fI\,FILE\/\fR [[\fI\,OPTION\/\fR] \fI\,FILE\/\fR]

.SH DESCRIPTION
difftime is a simple utility to calculate and print out the delta time from 
two files or a file and now. It's usefull to make a gate in a script, based on time.

The output is always a number. It can be signed with a '-' before the digits. 
By default there isn't a new line character.

The first argument must be a file. With no second argument, it prints out the
time differnce between the last modification time of the file and now. 

The output is in seconds by default.

Passing the second argument, the difference is between the modification time of
the second and the first. See OPTIONS to change this behaviour.

.SH OPTIONS
.TP
\fB\-a\fR, \fB\-\-atime\fR \fI\,FILE\/\fR
compare last access time
.TP
\fB\-c\fR, \fB\-\-ctime\fR \fI\,FILE\/\fR
compare change time
.TP
\fB\-m\fR, \fB\-\-mtime\fR \fI\,FILE\/\fR
compare modification time
.TP
\fB\-f\fR, \fB\-\-field\fR \fI\,field\/\fR
diff only a part of the timestamp. 'Y' year, 'M' month, 'D' day, 'h' hour, 'm' minute, 's' second.
.TP
\fB\-A\fR, \fB\-\-absolute\fR
output without sign
.TP
\fB\-D\fR, \fB\-\-days\fR
output in days
.TP
\fB\-H\fR, \fB\-\-hours\fR             
output in hours
.TP
\fB\-M\fR, \fB\-\-minutes\fR            
output in minutes
.TP
\fB\-P\fR, \fB\-\-print\fR             
add newline character after the number
.TP
\fB\-h\fR, \fB\-\-help\fR               
display this help message 
.TP
\fB\-v\fR, \fB\-\-version\fR            
display the software version 
.SH "EXIT STATUS"
With no arguments, it fail with exit code not zero.

If the file doesn't exists it fail with exit code not zero.
.SH BUGS
Bugs can be reported at 
https://github.com/ognibit/difftime/issues
.SH EXAMPLES
Show how old is a file (from now) in minutes
.
.IP
.EX
difftime -A -P -M /var/spool/myspool 
.EE
.
.P
Show time difference from files in days
.
.IP
.EX
difftime -P -D /var/spool/myspool myfile
.EE
.
.P
Show the time difference in seconds between the access time of one file and the
change time of the other
.
.IP
.EX
difftime -P -A -a /var/spool/base -c /var/spool/gate
.EE
.
.P
A simple way to use difftime into a script to do some operation only after 10
seconds to the previous effective execution.
.
.IP
.EX
#!/bin/bash

SPOOL_FILE="spool"
# time in seconds
TIME_GATE=10

DT=$(./difftime -A $SPOOL_FILE 2> /dev/null ) || DT=$TIME_GATE

[ $DT -lt $TIME_GATE ] && exit 1

echo "Do the stuff"

# reset the spool time
touch $SPOOL_FILE
.EE
.
.P
Execute a command once a day (like anacron)
.
.IP
.EX
DT=$(./difftime -A -f D $SPOOL_FILE 2> /dev/null ) || DT=1
[ $DT -eq 0 ] && exit 1
.EE
.SH AUTHORS
Omar Rampado <https://github.com/ognibit>
.SH "SEE ALSO"
stat(1)
difftime(3)


difftime.c
#include <errno.h>
#include <getopt.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h> 
#include <string.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

const char * help_msg = "Help \n\
version: %s \n\
usage: %s \
[-a|--atime] \
[-c|--ctime] \
[-m|--mtime] \
[-f]--field YMDhms] \
[-A|--absolute] \
[-D|--days] \
[-H|--hours] \
[-M|--minutes] \
[-P|--print] \
[-h|--help] \
[-v|--version] \
file1 [file2] \
\n\
\n\
difftime is a simple utility to calculate and print out the delta time from \n\
two files or a file and now. It's usefull to make a gate in a script, based on time.   \n\
\n\
Options: \n\
    -a, --atime file         compare last access time\n\
    -c, --ctime file         compare change time\n\
    -m, --mtime file         compare modification time\n\
    -f, --field Y|M|D|h|m|s  diff only a part of the date \n\
    -A, --absolute           output without sign\n\
    -D, --days               output in days\n\
    -H, --hours              output in hours\n\
    -M, --minutes            output in minutes\n\
    -P, --print              add newline character after the number\n\
    -h, --help               display this help message \n\
    -v, --version            display the software version \n\
\n\
Example: \n\
    difftime -P /var/spool/older /var/spool/newer\n\
    difftime -P /var/spool/file \n\
    difftime -P -D -c /var/spool/file -a file2 \n\
    difftime -P -f D -c /var/spool/file -a file2 \n\
";

static struct option const long_options[] = {
    {"atime",          required_argument, NULL, 'a'},
    {"ctime",          required_argument, NULL, 'c'},
    {"mtime",          required_argument, NULL, 'm'},
    {"field",          required_argument, NULL, 'f'},
    {"days",           no_argument, NULL, 'D'},
    {"hours",          no_argument, NULL, 'H'},
    {"minutes",        no_argument, NULL, 'M'},
    {"absolute",       no_argument, NULL, 'A'},
    {"print",          no_argument, NULL, 'P'},
    {"help",           no_argument, NULL, 'h'},
    {"version",        no_argument, NULL, 'v'},
    {NULL, 0, NULL, 0}
};

enum TimeField {
    ATIME,
    CTIME,
    MTIME
};

enum OutputUnit {
    SECONDS,
    MINUTES,
    HOURS,
    DAYS
};

struct {
    bool print;
    bool absolute;
    enum TimeField from_time;
    enum TimeField to_time;
    enum OutputUnit output_unit;
    char field;
} config;

void config_init(void);

#include "config.h"

void print_help(char *name);
void print_version(char *name);

void die(const char *prefix, const char *msg);

/*
 * time difference using config
 */
time_t timediff(time_t a, time_t b);

/* Get the time from st according to tf */
time_t extract_time(struct stat *st, enum TimeField tf);

/* convenient functions */
void config_time(char *arg, char **f, char **t, enum TimeField tm);
int config_field(char *arg);


int main(int argc, char *argv[])
{
    time_t tfrom, tto, tdiff; 
    char *ffrom, *fto; /* filenames */
    struct stat st;
    int c;
    char fmt[10];

    config_init();

    /* manage options */
    ffrom = NULL;
    fto = NULL;
    while ((c = getopt_long(argc, argv, "a:c:m:f:ADHMPhv", long_options, NULL)) > 0){
        switch(c){
        case 'a': /* atime */
            config_time(optarg, &ffrom, &fto, ATIME);
            break;
        case 'c': /* ctime */
            config_time(optarg, &ffrom, &fto, CTIME);
            break;
        case 'm': /* mtime */
            config_time(optarg, &ffrom, &fto, MTIME);
            break;
        case 'f': /* field */
            if (config_field(optarg) < 0){
                fprintf(stderr, "invalid field: %s\n", optarg);
                exit(EXIT_FAILURE);
            }
            break;
        case 'h': /* help */
            print_help(argv[0]);
            exit(EXIT_SUCCESS); 
            break;
        case 'v': /* version */
            print_version(argv[0]);
            exit(EXIT_SUCCESS); 
            break;
        case 'A': /* absolute */
            config.absolute = true;
            break;
        case 'D': /* days */
            config.output_unit = DAYS;
            break;
        case 'H': /* hours */
            config.output_unit = HOURS;
            break;
        case 'M': /* minutes */
            config.output_unit = MINUTES;
            break;
        case 'P': /* print */
            config.print = true;
            break;
        default:
            fprintf(stderr, "%s --help for usage\n", argv[0]);
            exit(EXIT_FAILURE);
        }
    }

    /* Extract the times */
    errno = 0;
   
    if (ffrom == NULL){
        if (optind >= argc)
            die("First file","filename required. See -h");
        ffrom = argv[optind++];
    }

    if (stat(ffrom, &st) < 0)
        die(ffrom, strerror(errno));

    tfrom = extract_time(&st, config.from_time);

    /* second argument can be set with option (like -c), wihout option
     * or not set (use time instead)
     */
    if (fto == NULL && optind < argc){ /* set without option */
        fto = argv[optind++];          /* get from argv and continue */
    }

    if (fto != NULL){ /* with filename */
        if (stat(fto, &st) < 0)
            die(fto,strerror(errno));

        tto = extract_time(&st, config.to_time);
    } else {   /* not provided, use the current time */
        if (time(&tto) < 0)
            die("Get current time",strerror(errno));
    }

    /* OUTPUT */

    if (config.print){
        strcpy(fmt, "%ld\n");
    } else {
        strcpy(fmt, "%ld");
    }

    tdiff = timediff(tto, tfrom);

    if (config.field == '\0'){
        /* apply the conversion only when diff whole time */
        switch (config.output_unit){
        case SECONDS:
            /* nothing to do */
            break;
        case MINUTES:
            tdiff /= 60; /* seconds in 1 minute */
            break;
        case HOURS:
            tdiff /= 3600; /* seconds in 1 hour */
            break;
        case DAYS:
            tdiff /= 86400; /* seconds in 1 day */
            break;
        }
    }

    printf(fmt, tdiff);

    return 0;
}

void print_help(char *name)
{
    fprintf(stderr, help_msg, VERSION, name);
}

void print_version(char *name)
{
    fprintf(stdout, "%s: %s\n", name, VERSION);
}

void die(const char *prefix, const char *msg)
{
    if (prefix){
        fprintf(stderr, "%s: ", prefix);
    } else {
        fprintf(stderr, "FATAL: ");
    }

    fprintf(stderr, "%s\n", msg);
    exit(EXIT_FAILURE);
}

time_t timediff(time_t a, time_t b)
{
    time_t d;
    struct tm *t;

    switch (config.field){
    case 's':
        t = gmtime(&a);
        d = t->tm_sec;
        t = gmtime(&b);
        d -= t->tm_sec;
        break;
    case 'm':
        t = gmtime(&a);
        d = t->tm_min;
        t = gmtime(&b);
        d -= t->tm_min;
        break;
    case 'h':
        t = gmtime(&a);
        d = t->tm_hour;
        t = gmtime(&b);
        d -= t->tm_hour;
        break;
    case 'D':
        t = gmtime(&a);
        d = t->tm_mday;
        t = gmtime(&b);
        d -= t->tm_mday;
        break;
    case 'M':
        t = gmtime(&a);
        d = t->tm_mon;
        t = gmtime(&b);
        d -= t->tm_mon;
        break;
    case 'Y':
        t = gmtime(&a);
        d = t->tm_year;
        t = gmtime(&b);
        d -= t->tm_year;
        break;
    default:
        d = difftime(a, b);
        break;
    }
    if (config.absolute)
        d = labs(d);

    return d;
}

time_t extract_time(struct stat *st, enum TimeField tf)
{
    time_t tm;

    switch (tf){
    case ATIME:
        tm = st->st_atime;
        break;
    case CTIME:
        tm = st->st_ctime;
        break;
    case MTIME:
        tm = st->st_mtime;
        break;
    }

    return tm;
}

void config_time(char *arg, char **f, char **t, enum TimeField tm)
{
    if (*f == NULL){
        config.from_time = tm;
        *f = arg;
    } else {
        config.to_time = tm;
        *t = arg;
    }
}

int config_field(char *arg)
{
    if (arg == NULL)
        return -1;

    if (strlen(arg) > 1)
        return -1;

    if (strstr("YMDhms", arg) == NULL)
        return -1;

    config.field = arg[0];
    return 0;
}
